### PR TITLE
Circuit breaker redirect

### PR DESCRIPTION
**Pull request by jkempster34**
### Trello card
https://trello.com/c/e9vBMLJE

### Context
When repeated API failures occur, circuit breaker middleware in the Ruby client will raise a `GetIntoTeachingApiClient::CircuitBrokenError`. This is an attempt to reduce the impact of any future service failures.

### Changes proposed in this pull request
- Redirect to service unavailable pages when there is a `GetIntoTeachingApiClient::CircuitBrokenError`

### Guidance to review
This PR is paired with https://github.com/DFE-Digital/get-into-teaching-api-ruby-client/pull/37

### CODE REVIEW

File: spec/components/service_unavailable_component_spec.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+require "rails_helper"
```
Comment (jkempster34) [inline spec/components/service_unavailable_component_spec.rb:1]: Not sure if this is useful?

File: spec/requests/circuit_breaker_spec.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  before do
+    api_constants = GetIntoTeachingApiClient.constants.select { |klass| klass.to_s.end_with?("Api") }
+    apis_to_mock = api_constants.map { |const| GetIntoTeachingApiClient.const_get(const) }
+    apis_to_mock.each do |api|
+      api.instance_methods(false).each do |method|
+        allow_any_instance_of(api).to receive(method).and_raise(error)
+      end
+    end
+  end
+
```
Comment (jkempster34) [inline spec/requests/circuit_breaker_spec.rb:15]: I wanted to parameterise these, but wasn't sure how to pass in the route helpers

File: spec/components/service_unavailable_component_spec.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+
+RSpec.describe ServiceUnavailableComponent, type: :component do
```
Comment (jkempster34) [inline spec/components/service_unavailable_component_spec.rb:3]: Not sure if this test is useful

Comment (peteryates) [reply spec/components/service_unavailable_component_spec.rb:3]: Makes sense to have it, even if it's minimal.

File: spec/requests/circuit_breaker_spec.rb (hunk 2 above)
Comment (peteryates) [reply spec/requests/circuit_breaker_spec.rb:15]: You should be able to do it by setting up outside of the RSpec block and looping through the scenarios you want to test. There's an example in the https://github.com/DFE-Digital/get-into-teaching-app/blob/master/spec/requests/redirects_spec.rb

File: app/controllers/concerns/circuit_breaker.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  end
+
+protected
+
+  def redirect_to_not_available
+    redirect_to(not_available_path)
+  end
+
+  def not_available_path
+    raise NotAvailablePathMissingError
```
Comment (peteryates) [inline app/controllers/concerns/circuit_breaker.rb:17]: Should this be returning a path rather than re-raising?

Comment (jkempster34) [reply app/controllers/concerns/circuit_breaker.rb:17]: I think this was meant as a way to ensure any class that extends the CircuitBreaker concern implements `not_available_path`

Comment (peteryates) [reply app/controllers/concerns/circuit_breaker.rb:17]: Ah, fair enough - maybe a tiny explanation in the error message could help here:

```ruby
raise NotAvailablePathMissingError, "#not_available_path hasn't been overridden"
```

**Review by peteryates — APPROVED** [review]
Everything here looks fine to me.

**Pull request merged**